engine/classes/Elgg/Widgets/CreateDefaultWidgetsHandler.php

```diff
@@ -0,0 +1,89 @@
+<?php
+
+namespace Elgg\Widgets;
+
+/**
+ * Creates default widgets
+ *
+ * @since 4.0
+ */
+class CreateDefaultWidgetsHandler {
+	
+	/**
+	 * This plugin hook handler is registered for events based on what kinds of
+	 * default widgets have been registered. See elgg_default_widgets_init() for
+	 * information on registering new default widget contexts.
+	 *
+	 * @param \Elgg\Event $event <event>, <entity_type>
+	 *
+	 * @return void
+	 */
+	public function __invoke(\Elgg\Event $event) {
+		static $processed_events = [];
+		
+		if (isset($processed_events["{$event->getName()}.{$event->getType()}"])) {
+			return;
+		}
+		// only create default widgets once per event
+		$processed_events["{$event->getName()}.{$event->getType()}"] = true;
+		
+		$default_widget_info = elgg_trigger_plugin_hook('get_list', 'default_widgets', null, []);
+		$entity = $event->getObject();
+		
+		if (empty($default_widget_info) || !$entity instanceof \ElggEntity) {
+			return;
+		}
+	
+		$type = $entity->getType();
+		$subtype = $entity->getSubtype();
+	
+		// event is already guaranteed by the hook registration.
+		// need to check subtype and type.
+		foreach ($default_widget_info as $info) {
+			if (elgg_extract('entity_type', $info) !== $type) {
+				continue;
+			}
+	
+			$entity_subtype = elgg_extract('entity_subtype', $info);
+			if ($entity_subtype !== ELGG_ENTITIES_ANY_VALUE && $entity_subtype !== $subtype) {
+				continue;
+			}
+	
+			// need to be able to access everything
+			elgg_push_context('create_default_widgets');
+	
+			elgg_call(ELGG_IGNORE_ACCESS, function () use ($entity, $info) {
+				// pull in by widget context with widget owners as the site
+				// not using elgg_get_widgets() because it sorts by columns and we don't care right now.
+				$widgets = elgg_get_entities([
+					'type' => 'object',
+					'subtype' => 'widget',
+					'owner_guid' => elgg_get_site_entity()->guid,
+					'private_setting_name' => 'context',
```

Comment:
why not a ``private_setting_name_value_pair``?

Reply:
jjust a copy and paste